product: mvp spec (#24)
aranya-project/aranya-docs: docs/aranya-mvp.md

> ---

Reviewer: This Front Matter block will need to be updated to something like this to show up in the site menu.
```
---
layout: page
title: Aranya Product MVP
permalink: "/aranya-product-mvp/"
---
```

taxonomy:
    category: docs
---

# Aranya Product MVP Spec

## Introduction

> A product specification for version 1 of the standalone Aranya daemon. The goal is to provide a 

Reviewer 2: This spec is for both the user library and daemon. The high-level API is mostly the API of the user library (although many of the API calls pass down into the daemon, some may not or could be modified slightly first).
```suggestion
A product specification for version 1 of the standalone Aranya daemon and user library. The goal is to provide a 
```

> commercial off the shelf solution for integrating Aranya. Customers should be able to download the Aranya daemon and user library, setup a team, and begin using Aranya.

Reviewer 2: ```suggestion
commercial-off-the-shelf solution for integrating Aranya. Customers should be able to download the Aranya daemon and user library, setup a team, and begin using Aranya.
```

Much of this document will serve as an extension to the [Aranya Beta spec](/docs/aranya-beta.md) that was implemented 
as version beta of the product. To capture all details related to the MVP in a single spot, 
relevant information from the beta spec has been carried into this document below. 

Primary Goals:

1. Provide a low friction solution customers can use to better secure their infrastructure.
2. Easily setup a team and onboard devices.
3. Implement a default policy that works well enough for a wide variety of situations that 
customers can relate to.

> 1. Provide stable APIs that allow devices and users to interact with Aranya.

Reviewer 2: Wonder if it's worth mentioning here that the APIs will be guaranteed to be backward compatible. Or if readers will know that's implied by "stable APIs".

> 2. Expose an API for point to point high performance encrypted communication using IPv4 for  

Reviewer 2: ```suggestion
2. Expose an API for point-to-point high performance encrypted communication using IPv4 for  
```

transport.

Secondary internal goals:

> 1. Design for the ability to swap out policies in a future version.

Reviewer 2: Is this for swapping the `policy.md` file out in the daemon at runtime? I'm not sure how useful this would be if the policy is not compatible with the user library APIs and client -> daemon APIs.
We currently can swap the policy out at build time but not runtime. And swapping the policy out could involve a lot of manual changes to the affected APIs.

Reviewer 3: @gknopf-aranya The idea is to initially allow custom policies that have the same interface but different implementation.

Reviewer 4: I was thinking something more along the lines of being able to substitute custom policies with new interfaces while being able to reuse as much of the daemon/client as possible. This functionality isnt planned for MVP or even slated for post-MVP, is it a longer term goal.

> 2. Design for future improvements, like additional data planes and configurable roles.

Reviewer 4: Looks like we currently support IPv6, will add a note to verify.

## Usage Requirements

Aranya is a decentralized message delivery platform with authorization built in. Below are some 
basic requirements for running this version of the product:

System Requirements:

> - We will target x86, ARM, and ARM64 running Linux.

Reviewer 2: Our open-source cicd pipeline doesn't support arm yet

Reviewer 2: It's a challenge because GitHub doesn't have arm cloud runners. So we'd most likely need to either use containers or figure out a way to secure our self-hosted arm runners. It's a whole project on its own that needs engineering time and devops support.

Reviewer 4: I added a note about this and put it on the list for MVP.

  - Mac
- We will assume there is IPv4 or IPv6 connectivity.

> TODO: measure the following values to estimate system requirements and include results in this spec.

Reviewer 2: Will these metrics be measured in the cicd pipeline?

Reviewer 3: I think we just need to measure them manually to get a ballpark of requirements.

Reviewer 4: Would be nice to have as part of the CICD pipeline, but I think a non-CI setup to calculate these is sufficient.

Reviewer 4: I added two tasks for the MVP: 

- Collect the current metrics and add to the spec
- Add the metrics to CI.

- Memory usage
- Disk requirements:
  - Storage amount
  - Storage device write speed
  - Storage device seek speed

TODO: include compile time and runtime requirements

## Architecture

There will be two subsystems as part of the product:

1. The daemon

> 2. The client library

Reviewer 2: The client library will be distributed as different artifacts for each supported application API. E.g. C API, Rust, etc.
We may wish to mention which languages we're planning to provide APIs for in this MVP spec.

Reviewer 2: We already have the C API, and there's a PR nearly ready with a Rust example app.

The daemon is a standalone process that runs Aranya and exposes the control plane API via IPC. The 

> daemon will periodically sync with peers and handle commands as they are synced. The current design 

Reviewer 2: We could mention here that peers need to be configured first before the daemon syncs with them. The way it's worded now makes it sound like the daemon will start syncing with peers automatically when it starts up.

does not require any disk-persisted effects, so on startup the daemon should start from a clean 
state (aside from loading accounts) every time it is started. This includes clearing any shared 
memory on startup to avoid collisions/stale data.

The client library exists inside the device's process and is used to interact with the daemon. The 
client library also includes a light wrapper around Quic Channels (and similarly for AFC), allowing 
the device to encrypt data and send messages to peers.

Both subsystems are limited to a single KeyBundle per process to limit the possibility of leaking 
device keys. This still enables the daemon to participate in multiple teams by using the same 

> keybundle for every team.

Reviewer 5: > Both subsystems are limited to a single KeyBundle per process to limit the possibility of leaking device keys. 

Having one KeyBundle per process definitely makes it easier to reason about. But it's not clear how it helps prevent us from leaking device keys. If we're going to keep this line then we need to explicitly state how it helps us prevent leakage.

>  This still enables the daemon to participate in multiple teams by using the same 
keybundle for every team.

I'm not exactly a fan of using the same KeyBundle across multiple times. It's only safe iff (a) the primitives are *exactly* the same, and (b) the policy uses the device keys in *exactly* the same manner. Case (a) isn't a /huge/ concern because the device keys are parameterized by the ciphersuite, but it's still possible for an `Engine` to be buggy. Case (b) is more difficult since it requires analyzing the policy.

The proper fix for this is to have a hierarchy like:

```
       IdentitySigningKey
        /                        \
   (signs)                 (signs)
      /                             \
KeyBundle1             KeyBundle2
```

And then add pk(IdentitySigningKey) to the graph alongside pk(IdentityKey), pk(EncryptionKey), etc.

Reviewer 5: /cc: @jdygert-spok @mooreoak since I guess this PR got moved here

Author: pre-mvp: add identity signing key for signing keys across different teams to associate the same entity with multiple key bundles

Reviewer 2: Is there a way to bind a key to a particular policy and crypto primitive implementation? It would be cool if we could reject keys that did not originate from matching policy and crypto primitives.

Reviewer 2: E.g. by signing the hash of the implementation

Reviewer 5: @gknopf-aranya this is already how it works

Reviewer 5: @gknopf-aranya  See this part of my comment:

> Case (a) isn't a /huge/ concern because the device keys are parameterized by the ciphersuite,

Reviewer 5: For example: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/aranya.rs#L154-L171

Reviewer 5: Every cryptographic operation ([signing data](0), [encrypting data](1), etc.) includes [the cipher suite identifiers](2) for *all* cryptographic algorithms used by the current policy. It mixes those identifiers into the cryptographic operation differently depending on the operation. This "binds" those cryptographic operations to that particular cipher suite. In other words, `Encrypt(data, CipherSuiteA) == Encrypt(data, CipherSuiteB)` iff `CipherSuiteA == CipherSuiteB`.

Secret keys are stored in a [`KeyStore`](3), which only operates on [wrapped* keys](4). [`Engine`](5)s are responsible for wrapping and unwrapping keys. It is the `Engine`'s responsibility to ensure that a wrapped key for algorithm A cannot be unwrapped as a key for algorithm B. The default `Engine` implementation does this by [binding the ciphertext](9) to the algorithm identifier *and* the key's unique ID. Other `Engine`s *should* do the same thing, but they might not. **This is one of my criticisms of sharing keys across teams.**

Public keys are stored in the fact database. They are [serialized in a format](8) that includes the cipher suite identifiers and key data type. Deserialization [should fail](10) if either are incorrect.

*: "Wrapped" means either encrypted and authenticated *or* stored in an HSM.

[0]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/aranya.rs#L154-L171
[1]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/groupkey.rs#L135-L159
[2]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/ciphersuite.rs#L72-L116
[3]: https://docs.rs/aranya-crypto/latest/aranya_crypto/keystore/trait.KeyStore.html
[4]: https://docs.rs/aranya-crypto/latest/aranya_crypto/engine/trait.WrappedKey.html
[5]: https://docs.rs/aranya-crypto/latest/aranya_crypto/engine/trait.Engine.html
[8]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/misc.rs#L404-L416
[9]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/default.rs#L99-L110
[10]: https://github.com/aranya-project/aranya-core/blob/3831ef2ab937ae4a0d28b8ad84aae65e527c05c0/crates/aranya-crypto/src/misc.rs#L362-L377

Reviewer 4: > @gknopf-aranya See this part of my comment:
> 
> > Case (a) isn't a /huge/ concern because the device keys are parameterized by the ciphersuite,

I am moving towards omitting the one-identity-per-subsystem requirement. I do think we will need additional thought into how we plan to expose this to the user. 

re (b), I am rewording it to better reflect what you posted above Eric: that we can use the same identity across teams using some sort of scheme to generate child/leaf keybundles from a root identity key. Any notes about the specific implementation of that in the spec are subject to change when we are working on it, as we might discover additional requirements.

I believe you mentioned last week @elagergren-spideroak  that this was something you wanted to include in MVP? Or at least, have some of the groundwork down for it.

Reviewer 2: When we create the issue to implement this portion of MVP, let's include a note to remember to update the MVP spec to reflect the implementation.

![daemon subsystems](/docs/images/new_daemon_arch_block.png)

TODO: update diagrams to remove AFC (https://github.com/aranya-project/aranya-docs/issues/6)
![daemon subsystems with additional detail](/docs/images/new_daemon_arch_block_detail.png)

For V1, we will initially design for the C API. The Rust client library is a bonus that we gain 
from the Rust -> C compilation. The C API should include autogenerated docs.

TODO (post-mvp): add support for other languages.

## Config

> On startup, the daemon requires the work directory as well as the path to the UDS socket that 

Reviewer 2: @ycarmy, @mooreoak UDS is an acronym we've been using internally but isn't widely used.
```suggestion
On startup, the daemon requires the work directory as well as the path to the unix domain socket that 
```

> should be created. The daemon working directory will be split using XDG, with namespaces and 

Reviewer 2: Since we're using XDG, we'll want to define what goes in `data`, `config`, and `state` locations:
https://specifications.freedesktop.org/basedir-spec/latest

Author: I believe in a previous discussion we concluded that this level of detail should be determined by whoever who will work on this feature, outlining the details in a separate spec (TODO captured in the last section of this doc).

prefixes for everything. Just the config path will be passed in.

> All other config values are provided when a client context is initialized and passed to the daemon 

Reviewer 2: This is how the daemon's config is implemented currently:
https://github.com/aranya-project/aranya/blob/main/crates/aranya-daemon/src/config.rs#L15
It contains more than just the unix domain socket path. E.g. the Aranya sync socket address, afc shm config.
Should we have the user library configure those later?

Reviewer 5: > Should we have the user library configure those later?

For the MVP the daemon should tell the client where to connect. See https://git.spideroak-inc.com/spideroak-inc/aranya-archive/issues/90

After the MVP we can consider adding RPC calls to update parameters like number of channels in the shared memory, etc.

over the UDS socket. This approach allows the device API to drive the configuration of the daemon, 
and can help reduce errors in config mismatch by minimizing the number of duplicate config values.

A config object will be used to instantiate a graph, i.e. in the action to perform the Init command 
and the API call for adding an existing graph (see CreateTeam and AddTeam APIs below). The config 
object will be versioned and extensible, and should not break API compatibility. 

> TODO (post-mvp): pass policy through Init command config object.

Reviewer 2: Is this tracked in a backlog or roadmap somewhere?

Author: Not yet. Once the MVP is complete, we will likely work on a plan for V2 of the product, where we will come back to this spec and pull out all the post-mvp tasks to establish where they should fit into the roadmap. This is what we did with the beta spec when planning for the MVP.

## Components

The system is made up of multiple components:

The "client API" is the top level component and contains local operations such as
enabling/disabling syncing or initializing a new device.

The Access Control Plane is the top level control plane, enabling IDAM operations
and other on-graph operations. The Access Control Plane is used to manage keys,
address assignment, roles, and labels as set out in the policy which is written in the MVP using 
[version 2.0 of the policy language](/docs/policy-v2.md).

> For the MVP, Aranya Quic Channels will be built to provide a simple API for sending and receiving 

Reviewer 2: Do we want to introduce the `AQC` acronym for Aranya QUIC Channels here?

Author: I don't know. I think some people had opinions against using acronyms. Will bring it up on our next call with the team.

Reviewer 4: I dont see this as blocking to resolve, I am going to add a note that it might be called AQC in case someone ever searches for AQC looking for it.

messages using a modified Quic transport. The Aranya Quic Channels contains its own control plane 
for control messages, as well as the main data plane for moving data between devices. This is 
similar to Aranya Fast Channels (AFC) from the beta version but with a different underlying 
transport (AFC uses TCP). See [Aranya Quic Channels API](#aranya-quic-channels-api).

Component structure:

- Local Client API: syncing, local device management
  - Access control plane (IDAM control plane): IDAM lifecycle
    - Quic Channels
      - Quic Channels control plane
      - Quic Channels data plane

>     - AFC (experimental)

Reviewer 2: Should we hide the AFC API behind an experimental flag so the user needs to enable AFC before using it?

Author: Maybe. I think that makes sense to do but I'm not sure what was originally discussed when deciding to make this "experimental". I'll bring it up on our next call.

Author: yes we should use a cargo rust feature for this

      - AFC control plane
      - AFC data plane
    - Message Broker (later...)
      - Broker control plane
      - Broker data plane
    - On graph messaging (later...)
      - graph messaging control plane (or implicit)
      - data plane (on graph messages)
    - ... additional planes in future versions

TODO: add table of comparison for AFC and Quic channels.

> ### Client APIs

Reviewer 5: we'll also want the `connect()` and other APIs used to init/close the connection with the daemon

Author: Do those APIs exist currently or do we plan on adding them for MVP? I'd need a bit more detail to add them into the spec correctly, like input/output parameters for each.

Reviewer 4: @elagergren-spideroak can you elaborate on this? it looks like for creating a client it automatically connects to the daemon. Is this in reference to data plane work?

The client APIs are local-only API endpoints that do not create commands on the graph. They are 
mostly used to manage the local state.

- `InitDevice() -> device_id` - init a device if not exists, generate keys, create the API 
instance, etc.
- `GetKeyBundle() -> keybundle` - returns the current device's public key bundle.
- `GetDeviceId() -> device_id` - returns the device's device ID.
- `AddTeam(team_id, config) -> bool` - add an existing team to the local device store. Not an 
Aranya action/command.
- `RemoveTeam(team_id) -> bool` - remove a team from the local device store. Not an Aranya action/
command.
- `SerializeKeyBundle(scheme, keybundle) -> bytes` - serialize a keybundle to a given format/
scheme. Ideally, this can be used to serialize to either human readable or machine readable formats.
- `DeserializeKeyBundle(scheme, bytes) -> keybundle` - desrialize a keybundle from a given format/
scheme.
- `SerializeId(scheme, id) -> bytes` - serialize an ID to a given format/scheme.

> - `DeserializeId(scheme, bytes) -> id` - deserialize an ID from a given format/scheme.

Reviewer 5: Remind me why we have these again?

Reviewer 4: I think they got copied over from an older spec. If we have improved the handling/ergonomics of serializing/deserializing IDs from C then we probably dont need to include them.

Reviewer 2: This was for using our Rust serialization to make it easier for a C application to pass IDs around during onboarding.

Author: Will bring this up on the next call to determine whether any improvements have been implemented (or should be), or if these should remain as they are in the spec.

Reviewer 4: I think we should keep these in (or something like them), the IDs in the C API are currently exposed as a struct with a field `uint8_t __for_size_only[64];` 

We pretty much always serialize the IDs as base58, so we need a way to expose that. Is there a better way we can do this?

Reviewer 3: I think we should

1. expose the bytes either through a method or directly
2. expose ser/deser for base58 only

There is no need to provide multiple schemas since it is really a primitive and not a composite type.

- `AwaitCommand(command_id)` - waits until command ID is present in the graph. For testing 

> purposes only.

Reviewer 5: > For testing purposes only.

Then we should not expose it in the public API.

Reviewer 3: Needs to be at least semi-public so we can use it from integration tests and demos. There's also the use case of when you can't yet open a channel or something and want to try it again later, but I don't think that's ironed out much especially with shift to quic channels. There could be further use cases but hard to envision...

Author: Will bring this up on the next call.

Author: This API is not just for testing, it's useful for regular operations with Aranya channels that need to wait for a command to be synced before creating channel. Can leave for post-mvp.

Reviewer 4: I added a note AwaitCommand can wait for MVP, I imagine IsPresent will be the same? From my perspective, AwaitCommand is just a small wrapper around IsPresent

Author: Yeah, IsPresent should probably be the same. Keep in mind that, while we are "okay" to wait with this post-MVP, the desire is to get it in pre-MVP but we understand that some sacrifices are needed to meet the timeline. However, communicating it this way in the spec can help ensure that an issue is created for this work already now but that an appropriate label is used to mark it as the next immediate thing that gets done once time permits.

- `IsPresent(command_id) -> bool` - looks for command ID in the graph.

#### Sync API

The IDAM control plane will be managed by the daemon process which will be accessed with the APIs 
provided in [IDAM Control Plane API](#idam-control-plane-api) via an IPC mechanism. The daemon will 
be responsible for syncing state with peers. To enable, the APIs below will be provided to add and 
remove peers to sync from.

Initially, UDS and shm will be used for IPC between the daemon and client library.

> - `AddSyncPeer(address, team_id, rate, config) -> bool` - add a peer to start syncing with at a specific rate. Syncing should support DNS resolution in the case a domain name is used. Syncs immediately the first time. The config object includes details for authorization (potentially authentication too) and options for syncing by pushing or pulling.

Reviewer 5: Perhaps `rate` should be part of `config`.

- `RemoveSyncPeer(address, team_id) -> bool` - remove a sync peer associated with the given address,
team_id.
- `SyncNow(address, team_id)` - Trigger an immediate sync for the peer.

>   - TODO: Restrict to already added peer?

Reviewer 2: I don't see a reason to restrict to an already added peer. Adding a sync peer means that the daemon will regularly sync with that peer at a specified rate. They already have permission to sync with any peer whenever they want, it's just making it easier to stay up-to-date. If we user wants to sync with a peer immediately, I think we should let them.

Reviewer 5: If we do *not* restrict `SyncNow` to already added peers then we need to add a `config` parameter.

Reviewer 2: There could be peers we'd only want to sync with once during initialization. So in those cases being able to sync with `SyncNow` without setting up a periodic sync interval could be beneficial.

Reviewer 5: `config` is for much more than periodic sync intervals. It'll eventually be used for authenticating the peer, etc. `SyncNow` can ignore a hypothetical `config.rate` parameter.

Reviewer 2: I meant being able to call `SyncNow` without calling `AddSyncPeer` first seems useful. That's a good point too.

Author: So it sounds like `SyncNow` should be restricted to already added peers, yes?

Author: Conclusion: don't restrict to already added peers but take in optional config that would be ephemerally used. Otherwise, existing config from AddSyncPeer will be used. If no config is given and none exists from AddSyncPeer then fail or we might have a default config. Implementation details are out of scope of this doc but remember to keep backward compatibility in mind.

> TODO: (post-mvp?) include static peer config (e.g., using JSON) so that fewer API calls will 

Reviewer 4: to be clear, this is suggesting some sort of bulk "import peers" functionality?

Author: I think that's the intended idea. @jdygert-spok?

Reviewer 3: Yes, makes it easier to set up lots of peers, or to recompile without a recompile.

We might have talked about this but I wonder if the peers should be persisted.

Author: This will be left for post-mvp. If there's time during MVP implementation then we can just do it.

be needed to set up

#### Onboarding API

**MVP**:

Easy to implement, key moving is done by integration.

1. Create Device (NewDevice)
2. Get Device Key (Current device KeyBundle (what you give to the admin))
3. Give KeyBundle to admin on team (integration problem)
4. Admin does AddDevice(device_key_bundle)
5. Get team_id from admin (integration detail)
6. Add team_id to client (AddTeam)
7. Sync with device on team (AddSyncPeer)

**Post-MVP**:

We will provide an additional optional onboarding API:

> TODO: we need a flow chart or just event sequence for these.

Reviewer 4: Are we still planning an update to onboarding for MVP?

Author: The update described in this section is not planned for MVP but still discussed for post-MVP.

- `CreateInvite(server address, team_id) -> Invite code` - create an invite code that will be sent 
to a device to onboard.
- `PollInvite(server address, Invite Code) -> Option<Device key>` - check the status of an invite  
code, returns the device ID of the device that joined using that code.
- `Join(server address, Invite Code, Device Key) -> Result<team_id>` - join a team using an invite  
code, returns the team_id of the team that was joined.

The goal of this onboarding API is to simplify the process of onboarding a user/device by providing 
an invite code instead of passing a KeyBundle.

#### Graph Querying APIs

FactDB queries over the current perspective of the graph should be possible through ephemeral 
commands in the policy that will return query results in their emitted effects.

TODO: List out query APIs against fact db
    - Role lookup: input device ID ; output assigned role(s)
    - KeyBundle lookup: input device ID ; output stored KeyBundle

>     - ...

Reviewer 4: Given that these all depend on a general way to access the factdb, we should probably focus more on the internal API/impl/capabilities we want to implement over a specific API at first.

- `RevokeNetworkName(team_id, device_id, net_identifier)` - disassociate a network address from a 
device.
- `CreateLabel(team_id, label)` - create a label
- `DeleteLabel(team_id, label)` - delete a label
- `AssignLabel(team_id, device_id, label)` - assign a label to a device so that it can be used for 
Quic Channels (or AFC)
- `RevokeLabel(team_id, device_id, label)` - revoke a label from a device
- `Finalization(team_id)` - create a truncation checkpoint of the graph. Any command received after 
a Finalization must be its descendant, otherwise, it will be dropped.
  - TODO: include details on minimal implementation?
    - (post-mvp?) add ability to propose change for after nth finalization

TODO: Custom roles actions (Client APIs for static policy or Control Plane APIs for dynamic policy?)
  - `DefineRole(role_name)` - define a new role type that can be assigned to devices
  - `ReplaceCommandRole(command_name, authoring_role)` - allow a particular role type to author the 
  specified command type
  - `ReplaceEntityClass(team_id, device_id, entity_class)` - replace the entity class associated to 
  a device
  - CLI to manage permissions, etc 

> ### Aranya Quic Channels API

Reviewer 2: Can we add a todo to this section to link to the quic channels spec?

- define/undefine channel labels
- assign/remove addresses/names for Aranya channels
- cannot send data on channel labels
- sync

`operator`

- add (new) / rm device in team
- assign member role to devices in team
- assign/revoke channel labels
- assign/remove addresses/names for channel
- cannot send data on channel labels
- sync

`member`

- use Aranya channel
- sync

`sync-role`

- sync

Notes:

- Devices can always remove and demote themselves
- Devices of equal role cannot remove or demote each other (hence, capabilities
of an owner can only be self-reduced)

## Documentation

API documentation must be provided for the client API covering the functions
and behavior of each API call. Most likely, this will take the form of a doxygen-like
web page. Developers can use this to look up language agnostic functions for operating
the client API. The documentation should also include tutorials and a quickstart
to get developers up and running with the product as soon as possible. Documentation
should also be provided for the daemon so that developers and sysadmins can
understand the requirements and operations of the daemon.

## Glossary

> - `AFC` - the library used to do high performance encryption using keys managed by `Aranya`.

Reviewer 2: Can add Aranya QUIC channels to the glossary

- `Aranya` - the main library that drives the control plane and policy execution.
- `daemon` - a long-lived process, typically running in the background, that
handles commands and keeps state.
- `device` - a computer, sometimes associated with a user but can also be independent.
In this model, we consider `devices` instead of users directly to accommodate
autonomous entities.
- `IPC` - inter-process communication.
- `policy` - an Aranya policy, containing the logic and rules of the system.
- `sync` - a request to synchronize the commands on the control plane. Syncs are currently
pull only, so the device that requests a sync receives commands from the requestee.
- `team` - a group of devices with an associated policy.

> - `UDS` - Unix Domain Sockets, used for communication between processes.

Reviewer 2: @ycarmy, @mooreoak we can remove this from the glossary if we're not using the `UDS` acronym

- `user` - a person who may operate a device.

### Verb pairs

- Add/Remove
- Assign/Revoke
- Create/Delete

## Additional Notes

> - All APIs use Result unless stated otherwise

Reviewer 2: Let's define `Result` in more detail. E.g. an object that contains a return value if the operation is successful, or an error if the operation has failed. Could reference a Rust `Result`, but this will look different from the C API.

<!-- TODO: Sort of exists, add to spec -->
- Need something similar to TLS SNI (ident name you are talking to inside TLS
connection, plus reverse SNI aka name you are talking from) for AFC messages.

> - TODO: C API naming prefix

Reviewer 5: I do not believe os.

> - TODO: C API naming prefix

Author: Will bring it up on the next call.

> - TODO: C API naming prefix

Author: 1. using namespaces for things like `create_bidi_channel` vs `create_uni_channel` etc.?
- just having uniform namespacing will help resolve the different cases here
2. should prefixes include receiver?
- yes
3. what if other parts are included, should prefix always be just `aranya`?
- yes, based on the library (e.g., "aranya"), receiver (e.g., "client"), and then the method name.

## TODO: Tasks needed to complete MVP updates

> -  Measure system requirements and record values in this spec.

Reviewer 4: I will group them similar to what we have for priorities (Urgent, High, Medium, Low).

Can take either DNS name, IPv4, or IPv6. Current implementation uses a bidi map, so we can reverse 
lookup.
- `UnsetAfcNetIdentifier(team_id, device_id, net_identifier)` - disassociate a network address from a device.
- `CreateAfcBidiChannel(team_id, peer_net_ident, label) -> channel_id` - create a bidirectional channel with the given peer.
- `DeleteAfcChannel(team_id, channel_id)` - delete a channel.
- `PollAfcData(timeout)` - blocks until new AFC data is available, or timeout elapsed
- `SendAfcData(channel_id, data)` - send data on the given channel.
- `ReceiveAfcData() -> (data, metadata)` - receive data from AFC.

### Graph Querying APIs

FactDB queries over the current perspective of the graph should be possible through ephemeral 
commands in the policy that will return query results in their emitted effects. These APIs
are likely to be moved to nice-to-have or Post-MVP, but are currently planned for MVP.

- `QueryRoleAssignment(device_id) -> Role`
- `QueryDeviceKeybundle(device_id) -> Keybundle`
- `QueryAfcNetworkId(device_id) -> network_str`
- `QueryAfcLabelAssignments(device_id) -> Vec<label>`

> - `QueryAfcLabelExists(device_id) -> Vec<label>`

Reviewer 2: @djohnson-spideroak shouldn't this be: 
`QueryAfcLabelExists(device_id, label) -> bool`
to check if the label exists?

Init command, contains the system's policy that defines the IDAM control plane for bootstrapping.

- `CreateTeam(owner) -> team_id` - initialize the graph, creating the team with the author as the 
owner. Includes policy for bootstrapping.
- `CloseTeam(team_id)` - close the team and stop all operations on the graph.
- `AddDevice(team_id, keybundle, common_name) -> device_id` - add a device to the team with the 
default role
- `RemoveDevice(team_id, device_id)` - remove a device from the team
- `AssignRole(team_id, device_id, role)` - assign a role to a device
- `RevokeRole(team_id, device_id, role)` - remove a role from a device
- `AssignNetworkName(team_id, device_id, net_identifier)` - associate a network address to a device 
for use with Quic Channels (or AFC). If the address already exists for this device, it is replaced 
with the new address. Capable of resolving addresses via DNS, required to be statically mapped to 
IPv4. For use with CreateChannel and receiving messages. Can take either DNS name or IPv4. 
  - TODO?: MVP would need reverse lookup. More work required on the address assignment, especially 

>   post-beta. Currently one name per device.

Author: A Quic channel ID will be used for this as we did with AFC.

Reviewer 4: Can you explain?

btw: current implementation supports IPv4+6, and also currently supports the reverse lookup. I am putting a note we should revisit the network name stuff later. About to push an updated version.

Author: From my understanding, we don't need to do reverse DNS lookup and the Quic channel ID is already being used to map the appropriate peer address, so this whole TODO might just need to be removed (cc: @elagergren-spideroak)

Reviewer 5: The current implementation neither needs nor supports reverse DNS lookups. Ditto for the future implementation.

- `RevokeNetworkName(team_id, device_id, net_identifier)` - disassociate a network address from a 
device.
- `CreateLabel(team_id, label)` - create a label
- `DeleteLabel(team_id, label)` - delete a label
- `AssignLabel(team_id, device_id, label)` - assign a label to a device so that it can be used for 
Quic Channels (or AFC)
- `RevokeLabel(team_id, device_id, label)` - revoke a label from a device
- `Finalization(team_id)` - create a truncation checkpoint of the graph. Any command received after 
a Finalization must be its descendant, otherwise, it will be dropped.

>   - TODO: include details on minimal implementation?

Author: Minimal implementation will just add a new command to graph with a finalization priority.

Then, support for finalization priority in braid algorithm so that it's higher than all other priority values and can never be equal to itself (i.e., can't compare 2 finalization command priority values). This can be post-mvp but would be nice to include in MVP if there's time.

Author: Even minimal finalization can be post-mvp

>     - (post-mvp?) add ability to propose change for after nth finalization

Author: This is related to sequencing finalizations so that when a change is proposed after nth finalization and current sequence number is n+1, then the change is allowed. This can be done in policy and could already be added in MVP.

Reviewer 4: Added a note about finalization being Post-MVP, and included a more detailed description of this. I created a new section at the end of the document for things we want to revisit, and I have started tagging things inline with `Post-MVP`

  - `ReplaceEntityClass(team_id, device_id, entity_class)` - replace the entity class associated to 
  a device
  - CLI to manage permissions, etc 

### Aranya Quic Channels API

TODO: add reference to Quic channels spec once ready

Note: The beta spec describes Aranya Fast Channels (AFC) which is replaced by Aranya Quic Channels 
for the MVP. AFC is instead viewed as an experimental feature of the product which might still be
preferred when running on embedded devices or when using a unidirectional transport. Refer to the 
beta spec and other existing documentation ([AFC](/docs/afc.md) and [AFC-Crypto](/docs/afc-crypto.md)) for more details on AFC. 

Just like AFC, the Quic Channels plane is split in two different sub-planes: the Quic channels 
control plane and the Quic channels data plane. The Quic channels control plane is responsible for 
any Aranya command or ephemeral commands, while the Quic channels data plane contains only data 
related APIs.

Aranya Quic channels use a modified Quic transport implementation that supports the ability to use 
custom cryptography and has latency-based congestion control (see (s2n-quic)[https://github.com/aws/s2n-quic]).

> TODO: not requiring TLS certs?

Author: nope

  
TODO: (post-mvp?) embedded devices that implement the Aranya library should still be able to sync 

> and use AFC to communicate with devices that have full product integrated.

Reviewer 4: Cleaned up the note about this, labeled Post-MVP, and added to the list.

#### Quic Channels Control Plane

> - `CreateQuicChannel(team_id, net_identifier, label) -> channel_id` - Open a channel to the dest  

Author: Should include head of creator's graph.

Reviewer 4: Do we need to include this as an argument or is it part of the functionality?

cc @elagergren-spideroak  @jdygert-spok  @gknopf-aranya

Reviewer 5: Off hand, I think it's something we should do internally

Reviewer 4: https://github.com/aranya-project/aranya-core/issues/89

with the given label. The device API transparently handles sending the ephemeral command to the 
peer. Channel keys are automatically rotated after a specific byte count.
- `DeleteQuicChannel(team_id, channel_id)` - Close a channel with the given ID. The device API   
transparently handles sending the ephemeral command to the peer.

#### Quic Channels Data Plane

> TODO: (post-mvp?) improving AFC polling in rust API

Author: Different mechanism for Quic channels will be used. Improving the API for AFC specifically is not part of MVP but if improvements are needed for Quic channel then they might happen for MVP - see Quic channels spec.

- `PollData(timeout) -> bool` - blocks with timeout, returns true if there is data on the channel 
to read using `ReadData`.
  - TODO: polling needs improvement, so don't copy over to Quic Channels

>   - TODO: need separate mechanisms for C and Rust

Author: Should be discussed under Quic channels spec. APIs will be very different.

Reviewer 4: @ycarmy How much of the QUIC channels API do we want to have here? Just the control plane? I imagine the APIs will be part of that spec, but I can also see a reason to include them here. For now, I am going to move this to the appendix, and give a link to the QUIC channels spec.

Reviewer 4: cc @benz-spideroak

Author: I think we might want to note all APIs here but in a summarized version, as we've done so far with the others, while providing a link to the Quic channels spec for more detail. I'm not sure yet whether there will be some lower level APIs for internal functionality that don't need to be indicated here but we can assess that once the Quic channels spec is complete before adding the APIs to this doc. 

If we expect the Quic channels spec will take much longer to complete, then we probably shouldn't block on merging this spec and instead just leave a TODO to make the updates in a separate PR once ready.

Reviewer 6: I don't have an opinion on what to include here. None of the APIs currently included here are the same in the QUIC channels spec. And many of the APIs here will be removed. Probably best to just link to the other spec for the time being.

Reviewer 4: I will remove the API endpoints from this section and link the spec.

Author: @benz-spideroak will the Quic channels spec include all the updated APIs for AFC (like for creating an AFC channel)? Feels like some of them might be more appropriate in this spec unless they are directly related to Quic channels too.

Reviewer 6: There will be nothing about AFC in the QUIC channels spec. AFC isn't changing.

Reviewer 2: Agree that the QUIC channels spec should include the top-level Aranya client APIs relevant to QUIC channels. This spec can reference that section of the QUIC channels spec. @benz-spideroak can you commit a work-in-progress section for that to the QUIC channels spec so we can update this PR with a link to that section and close this discussion out.

Reviewer 4: I am going to mark this as resolved, and leave a link to this issue here to update the MVP spec after the QUIC spec is further along:

https://github.com/aranya-project/aranya-docs/issues/66

- `ReadData(bytes_buffer) -> Result<(len_written, remote_net_identifier, label, channel_id)>` - 
read the channel data, returning the plaintext, sender, and label (or an error). Returns one 
channel message at a time.
- `SendData(bytes, channel_id, timeout) -> bool` - encrypts and sends the data to the given channel 
with a timeout. This call is blocking until the timeout is complete.

TODO: (post-mvp?)
For the initial implementation in the beta, AFC control messages should be handled
transparently by the client library. In the future, control messages can be passed

> to the device to be manually forwarded to the daemon using a different API.

Author: Post-mvp we will add a separate API for manually handling control messages in Quic channels / AFC. So nothing needed here for MVP.

to get developers up and running with the product as soon as possible. Documentation
should also be provided for the daemon so that developers and sysadmins can
understand the requirements and operations of the daemon.

## Glossary

- `AFC` - the library used to do high performance encryption using keys managed by `Aranya`.
- `Aranya` - the main library that drives the control plane and policy execution.
- `daemon` - a long-lived process, typically running in the background, that
handles commands and keeps state.
- `device` - a computer, sometimes associated with a user but can also be independent.
In this model, we consider `devices` instead of users directly to accommodate
autonomous entities.
- `IPC` - inter-process communication.
- `policy` - an Aranya policy, containing the logic and rules of the system.
- `sync` - a request to synchronize the commands on the control plane. Syncs are currently
pull only, so the device that requests a sync receives commands from the requestee.
- `team` - a group of devices with an associated policy.
- `user` - a person who may operate a device.

### Verb pairs

- Add/Remove
- Assign/Revoke
- Create/Delete

> ## Additional Notes

Reviewer 4: Is this something more specific to the rust implementation? It might be worth opening a ticket there instead.

> ## Additional Notes

Reviewer 2: True, that is implementation specific. @nikki-aranya once the issue is created, could you add a link to it here so we can close out this conversation?

- All APIs use Result unless stated otherwise
<!-- TODO: Sort of exists, add to spec -->
- Need something similar to TLS SNI (ident name you are talking to inside TLS

> connection, plus reverse SNI aka name you are talking from) for AFC messages.

Author: Not needed for Quic channels and already part of AFC. Remove from spec.

for control messages, as well as the main data plane for moving data between devices. This is 
similar to Aranya Fast Channels (AFC) from the beta version but with a different underlying 
transport (AFC uses TCP). See [Aranya Quic Channels API](#aranya-quic-channels-api).

Component structure:

- Local Client API: syncing, local device management
  - Access control plane (IDAM control plane): IDAM lifecycle
    - Quic Channels
      - Quic Channels control plane
      - Quic Channels data plane
    - AFC (build flag required to enable API)
      - AFC control plane
      - AFC data plane
    - ... additional planes in future versions

Rust features will be used for some features like the raw AFC interface. By default, AFC will not be included in the
user facing API unless a specific build flag is present. The goal of this choice is to better signal which APIs are
best suited for common use.

### Client APIs

The client APIs are local-only API endpoints that do not create commands on the graph. They are 
mostly used to manage the local state. Depending on the language, the endpoints may be a different
format that is more idiomatic to that language such as snake_case for C.

> - `InitDevice() -> device_id` - init a device if not exists, generate keys, create the API 

Reviewer 4: from what I can tell, we don't currently have an InitDevice endpoint, it looks like we do this when initializing a client.

Reviewer 2: @djohnson-spideroak from what I can tell, initialization ended up being implemented in the `connect` API call which connects to the daemon and initializes a few things.
```
    #[instrument(skip_all, fields(?daemon_sock, ?afc_shm_path, max_chans))]
    pub async fn connect<A>(
        daemon_sock: &Path,
        afc_shm_path: &Path,
        max_chans: usize,
        afc_listen_addr: A,
    ) -> Result<Self>
```
cc: @jdygert-spok

Reviewer 3: IIRC it is really done when the daemon starts up. Those args are all just for the client's connections.

Reviewer 2: The daemon is supposed to do the minimum amount of initialization on startup. E.g. setting up the unix domain socket API endpoint to receive data from the Aranya client. After that, additional configuration is supposed to occur by subsequent API calls from the Aranya client. As an example, right now the shared memory is configured in both the Aranya client and daemon. Ideally, that should be configured first in the Aranya client, then passed down into the daemon through the client <-> daemon unix domain socket API.

The reason to pass configuration info to the daemon through the API is so the configuration info only needs to be maintained in a single place.

Reviewer 4: I was looking at the C API for `InitDevice`. The notes on what the daemon should configure on startup vs on receiving a connection is represented in the Config section.

Author: Per this discussion, and an earlier comment by Eric, it seems that there might be some APIs still missing from the spec, like `connect`. If `InitDevice` does not exist, nor is planned to exist, then it should be removed from the spec.

- We will assume there is IPv4 or IPv6 connectivity.

Measure the following values to estimate system requirements and include results in this spec.
- Memory usage
- Disk requirements:
  - Storage amount
  - Storage device write speed
  - Storage device seek speed

For a full and up to date list, start with this issue: https://github.com/aranya-project/aranya/issues/62

## Architecture

There will be two subsystems as part of the product:

1. The daemon
2. The client library

The daemon is a standalone process that runs an Aranya instance and exposes the control plane API via IPC. 
It handles setting up all the dependencies that Aranya needs including storage, policy, and network communication.
The daemon will periodically sync with registered peers and handle commands as they are synced. The current design 
does not require that _effects_ are stored on disk. Other implementations have required
effects be stored on disk in case the program is terminated after the effect has been emitted, but before
it has been processed by the daemon or user application. This extends to the shared memory; there is no 

> requirement it be preserved between program restarts.

Reviewer 7: Make a note about this somewhere, since we have measuring system requirements on the TODO list?

For MVP we can lay the groundwork for the multi-keybundle approach, and better management of identities
can be added post-MVP or as time allows for MVP.

![daemon subsystems](/docs/images/new_daemon_arch.png)

![daemon subsystems with additional daemon](/docs/images/new_daemon_arch_detail.png)

## Config

On startup, the daemon requires a path to the working directory. That directory contains a configuration
file with the path to the unix domain socket that should be created and other values. The daemon working
directory will structured to be compatible with the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/latest/). Adhering to the XDG Base Directory Specification will increase the interoperability of Aranya
across multiple system configurations.

All other config values are provided when a client context is initialized and passed to the daemon  
over the Unix Domain Socket. This approach allows the device API to drive the configuration of the 
daemon, and can help reduce errors in config mismatch by minimizing the number of duplicate config 

> values.

Reviewer 3: Doing some daemon config from the client opens up questions about partially initialized daemons and what to do when trying to initialize something already initialized.

Reviewer 4: I think this is something we will mainly handle in implementation. I will add a note about this though.

format that is more idiomatic to that language such as snake_case for C.

- `InitDevice() -> device_id` - init a device if not exists, generate keys, create the API 
instance, etc.
- `GetKeyBundle() -> keybundle` - returns the current device's public key bundle.
- `GetDeviceId() -> device_id` - returns the device's device ID.
- `AddTeam(team_id, config) -> bool` - add an existing team to the local device store. Not an 
Aranya action/command.
- `RemoveTeam(team_id) -> bool` - remove a team from the local device store. Not an Aranya action/
command.
- `SerializeKeyBundle(scheme, keybundle) -> bytes` - serialize a keybundle to a given format/
scheme. Ideally, this can be used to serialize to either human readable or machine readable formats.
- `DeserializeKeyBundle(scheme, bytes) -> keybundle` - desrialize a keybundle from a given format/
scheme.
- `SerializeId(id) -> bytes` - serialize an ID to the standard base58 format. https://github.com/aranya-project/aranya-docs/pull/24/files#r1915516900
- `DeserializeId(bytes) -> id` - deserialize an ID from the standard base58 format. https://github.com/aranya-project/aranya-docs/pull/24/files#r1915516900

> - `IsPresent(command_id, max_cut) -> bool` - looks for command ID in the graph.

Author: Did you intentionally leave this in and take `AwaitCommand` out?

Author: @djohnson-spideroak ^

Reviewer 4: Did not intentionally leave it in, do we think it should be included in this version, or do we want to also push it to post MVP? I think it will probably get pushed

config is provided, use that. If no config arg is provided, fallback to the config used when the peer
was added via AddSyncPeer. If the peer was not added, use a default config or error. https://github.com/aranya-project/aranya-docs/pull/24/files#r1917188406

#### Onboarding API

**MVP**:

Easy to implement, key moving is done by integration.

1. Create Device (NewDevice)
2. Get Device Key (Current device KeyBundle (what you give to the admin))
3. Give KeyBundle to admin on team (integration problem)
4. Admin does AddDevice(device_key_bundle)
5. Get team_id from admin (integration detail)
6. Add team_id to client (AddTeam)
7. Sync with device on team (AddSyncPeer)

#### Graph Querying APIs

FactDB queries over the current perspective of the graph should be possible through ephemeral 
commands in the policy that will return query results in their emitted effects. These APIs
are likely to be moved to nice-to-have or Post-MVP, but are currently planned for MVP.

- `QueryFactRole(user_id) -> Role`
- `QueryFactKeybundle(user_id) -> Keybundle`
- `QueryFactNetworkId(user_id) -> network_str`
- `QueryFactLabelAssignments(user_id) -> Vec<label>`

> - `QueryFactLabelExists(user_id) -> Vec<label>`

Author: change "user" to "device"

### IDAM Control Plane API

The IDAM control plane is for managing identity and authorization by interacting with the graph. 
Each endpoint creates one or more commands on the graph. The first command in the graph, aka the 
Init command, contains the system's policy that defines the IDAM control plane for bootstrapping.

- `CreateTeam(owner_keybundle) -> team_id` - initialize the graph, creating the team with the author as the 
owner. Includes policy for bootstrapping.
- `CloseTeam(team_id)` - close the team and stop all operations on the graph.
- `AddDeviceToTeam(team_id, keybundle, common_name) -> device_id` - add a device to the team with the 
default role
- `RemoveDeviceFromTeam(team_id, device_id)` - remove a device from the team
- `AssignRole(team_id, device_id, role)` - assign a role to a device
- `RevokeRole(team_id, device_id, role)` - remove a role from a device

> - `SetNetworkName(team_id, device_id, net_identifier)` - associate a network address to a device 
> for use with Quic Channels (or AFC). If the address already exists for this device, it is replaced 
> with the new address. Capable of resolving addresses via DNS. For use with CreateChannel and receiving
> messages. Can take either DNS name, IPv4, or IPv6. Current implementation uses a bidi map, so we can
> reverse lookup.
> - `UnsetNetworkName(team_id, device_id, net_identifier)` - disassociate a network address from a 
> device.

Author: Sounds like AQC won't be using these network identifiers (cc: @elagergren-spideroak @benz-spideroak), so I wonder if we want the APIs to specify that they're for AFC?
```suggestion
- `SetAfcNetworkName(team_id, device_id, net_identifier)` - associate a network address to a device 
for use with Quic Channels (or AFC). If the address already exists for this device, it is replaced 
with the new address. Capable of resolving addresses via DNS. For use with CreateChannel and receiving
messages. Can take either DNS name, IPv4, or IPv6. Current implementation uses a bidi map, so we can
reverse lookup.
- `UnsetAfcNetworkName(team_id, device_id, net_identifier)` - disassociate a network address from a 
device.
```

Author: Implemented this change.

Aranya is a decentralized message delivery platform with authorization built in. Below are some 
basic requirements for running this version of the product:

System Requirements:

- We will target x86, ARM, and ARM64 running Linux.
  - Must run on Mac for development reasons.
  - ARM (arm32) is in progress, additional work is required to add it to CI. https://github.com/aranya-project/aranya-docs/pull/24/files#r1917143524
- We will assume there is IPv4 or IPv6 connectivity.

TODO: measure the following values to estimate system requirements and include results in this spec.
- Memory usage
- Disk requirements:
  - Storage amount
  - Storage device write speed
  - Storage device seek speed

TODO: include compile time and runtime requirements

## Architecture

There will be two subsystems as part of the product:

1. The daemon
2. The client library

The daemon is a standalone process that runs an instance Aranya and exposes the control plane API via IPC. 
It handles setting up all the dependencies that Aranya needs including storage, policy, and network communication.
The daemon will periodically sync with registered peers and handle commands as they are synced. The current design 

> does not require any disk-persisted effects, so on startup the daemon should start from a clean 

Reviewer 4: It appears this is referencing the effects only. In other integrations we have done, we had to build an effect queue and persist them to disk to be resilient to restarts. We had talked about this in a rust call awhile ago, we looked at the policy and saw that we dont have any effects that would need to be persisted and saved like that. I agree this should be clarified though.

[version 2.0 of the policy language](/docs/policy-v2.md).

For the MVP, Aranya Quic Channels (maybe called AQC in the future) will be built to provide a simple API for sending and receiving 
messages using a modified Quic transport. The Aranya Quic Channels contains its own control plane 
for control messages, as well as the main data plane for moving data between devices. This is 
similar to Aranya Fast Channels (AFC) from the beta version but with a different underlying 
transport (AFC uses TCP). See [Aranya Quic Channels API](#aranya-quic-channels-api).

Component structure:

- Local Client API: syncing, local device management
  - Access control plane (IDAM control plane): IDAM lifecycle
    - Quic Channels
      - Quic Channels control plane
      - Quic Channels data plane
    - AFC (experimental flag required)
      - AFC control plane
      - AFC data plane
    - Message Broker (later...)
      - Broker control plane
      - Broker data plane
    - On graph messaging (later...)
      - graph messaging control plane (or implicit)
      - data plane (on graph messages)
    - ... additional planes in future versions

> Rust features will be used for some features like the raw AFC interface. The goal of this choice is to better signal which APIs are best suited for public use. 

Reviewer 4: I will add it to the spec, but from what I understand of the discussions we had, AFC will be omitted from the public API by default, with an option to enable it when building via a flag during compilation.

format that is more idiomatic to that language such as snake_case for C.

- `InitDevice() -> device_id` - init a device if not exists, generate keys, create the API 
instance, etc.
- `GetKeyBundle() -> keybundle` - returns the current device's public key bundle.
- `GetDeviceId() -> device_id` - returns the device's device ID.
- `AddTeam(team_id, config) -> bool` - add an existing team to the local device store. Not an 
Aranya action/command.
- `RemoveTeam(team_id) -> bool` - remove a team from the local device store. Not an Aranya action/
command.
- `SerializeKeyBundle(scheme, keybundle) -> bytes` - serialize a keybundle to a given format/
scheme. Ideally, this can be used to serialize to either human readable or machine readable formats.
- `DeserializeKeyBundle(scheme, bytes) -> keybundle` - desrialize a keybundle from a given format/
scheme.
- `SerializeId(id) -> bytes` - serialize an ID to the standard base58 format. https://github.com/aranya-project/aranya-docs/pull/24/files#r1915516900
- `DeserializeId(bytes) -> id` - deserialize an ID from the standard base58 format. https://github.com/aranya-project/aranya-docs/pull/24/files#r1915516900

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 4: re use case: yes, that is correct. iirc we have done this before, I dont remember where though. Might also be useful if we include the command ID in handshakes/protocol headers when trying to initiate a connection though one of our data planes (if we arent doing that already). If we include the command ID, the receiving peer can AwaitCommand on the channel creation command if they dont already have it.

re MVP stuff: I think I am going to move anything Post-MVP into an appendix. Should help us ensure the scope of the main document matches what we want to achieve, and we still get to document planned/wanted features for the future. Would that address your concerns on the scope of the Post-MVP stuff?

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 5: >  we include the command ID in handshakes/protocol headers when trying to initiate a connection though one of our data planes (if we arent doing that already). 

Could you please give an example? I do not understand what you mean here.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 8: @elagergren-spideroak the message that dose the key agrement for quic channels should include the head of the graph it was generated on. This lets the receiver know if they are out of sync with the sender. The is useful if the auth fails on graph but the receiver dose not have the command in the head from the sender's graph. They would then initiasheat a sync with the sender and use the wait for the head command id in the open channel command before trying to reprocess it.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 5: Right, right, thanks.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 4: Would the head of the graph that gets sent over in the key agreement be authenticated/verfied? Could introduce a potential DOS issue if they send a bogus command ID and force the receiver to await on a command that will never be received.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 5: > Would the head of the graph that gets sent over in the key agreement be authenticated/verfied?

It's 

> Could introduce a potential DOS issue if they send a bogus command ID and force the receiver to await on a command that will never be received.

Yes, but it shouldn't be any worse than regular syncing. This is essentially what should occur:

- S sends create channel command to R
- R tries to add the command from S to the graph, but policy rejects it
- R notices the heads are different and assumes that might be the reason why it failed
- R syncs with S (*) and requests the missing portion of the graph (**)

S could delay syncing, or send parts of the graph that R already has. But it could already do that.

One difference is that while syncing happens in the background, `WaitForCommand` ostensibly blocks (the client library). But we don't have to implement it like that. We could implement something like:

```rust
impl Client {
    async fn wait_for_command(&self, id: Id) -> Result<WaitingForCommand, Error>;
}
struct WaitingForCommand { ... }
impl Future for WaitingForCommand { ... }
```

`wait_for_command` tells the daemon to notify the client when a particular command is received. The `WaitingForCommand` future is completed when the daemon notifies the client. The caller can then go do other work and periodically check the `WaitingForCommand` future, or even time out if it's taking too long.

*: Or somebody else.
**: I don't believe we currently have a way to say, "give me exactly this portion of the graph," though.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 5: > Would the head of the graph that gets sent over in the key agreement be authenticated/verfied?

Off the top of my head: sorta, kinda, not really.

You'd need the chain of commands from the head back to a common ancestor, which would (a) be prohibitively large for the create channel command and (b) would require figuring out a common ancestor between two graphs. You could _sorta_ get away with sending the chain of (author, signature, hash) tuples, but that requires all authors to be in the graph at that common ancestor. Worse, it totally ignores the policy in those commands. And it's still a lot of data.

The create channel command from the peer _is_ signed, so an adversary can't trick you. But a poorly behaved peer can still trick you.

We _might_ be able to come up with Really Cool Proofs for this sort of thing. But I don't think we've researched it much yet.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 8: We should make sure we have what we need it network packets for MVP even it we don't use it quite yet so as to allow us to implement this in the future as a non breaking change.

> - `AwaitCommand(command_id)` - waits until command ID is present in the graph. (Post-MVP)

Reviewer 4: added "-  Update key agreement commands to have a field for the sender's graph head ID. https://github.com/aranya-project/aranya-docs/pull/24#discussion_r1937642908" to the TODOs. I think this thread is resolved.

### Low
-  Set up CI to measure resource usage
-  Implement FactDB query APIs
-  CLI for managing permissions
-  Update default policy to have one role type authoring each command

### Nice to have

Anything moved to Post-MVP but would be nice to get into the MVP if we have resources.

-  Implement `Finalization`
-  Implement `AwaitCommand` and `IsPresent`.

### Uncategorized

-  Update spec diagrams.
-  Outline and apply daemon working directory changes.
  -  Establish best terminology for reference (e.g. "device store"?).
-  Improve polling mechanism for Quic channels

## Appendix B: Naming Schemes

The C API will use the following naming scheme:
```
(libraryPrefix)_((receiver)_)?(functionName)
```

Where 
- libraryPrefix is a global prefix for the whole library (ex: Aranya)
- receiver is a thing we are calling a function on (see below for example) 
- functionName is the name of the function to call.

For example, an API endpoint with the name `FooClientBar` breaks down as follows:

- `Foo` is the library.
- `Client` is the receiver.
- `Bar` is the function name.

`FooClientBar` calls the function `Bar` on an instance of `Client` (passed as an argument)

> provided by the library `Foo`.

Author: Are we already doing this or do we need a TODO in Appendix A to update existing APIs?

Reviewer 4: In some places we follow this, but in most of the C API it appears we dont. We have `aranya_client_cleanup` which matches, and others like `aranya_assign_role_ext`. So if we want to be consistent, we are going to have to rename the functions (or make AranyaClient an exception?).

Author: Okay so let's make sure to capture the work in a TODO in the appendix

### Additional Notes

- All Rust APIs use Result unless stated otherwise
- C APIs return the AranyaError type, and use parameters for return values unless
some other pattern is required.

## Appendix D: Glossary

- `AFC` - the library used to do high performance encryption using keys managed by `Aranya`.
- `Aranya Quic Channels (maybe AQC)` - An integration of Aranya with QUIC to provide a secure 
and integrated transport. See QUIC channels spec. TODO(declan): link
- `Aranya` - the main library that drives the control plane and policy execution.
- `daemon` - a long-lived process, typically running in the background, that
handles commands and keeps state.
- `device` - a computer, sometimes associated with a user but can also be independent.
In this model, we consider `devices` instead of users directly to accommodate
autonomous entities.
- `IPC` - inter-process communication.
- `policy` - an Aranya policy, containing the logic and rules of the system.
- `sync` - a request to synchronize the commands on the control plane. Syncs are currently
pull only, so the device that requests a sync receives commands from the requestee.
- `team` - a group of devices with an associated policy.

> - `user` - a person who may operate a device.

Author: Feels like these should be moved up, before the TODOs, as they contain relevant information to the core content of this spec.

Author: Sure that would be fine too I suppose.

the device to encrypt data and send messages to peers. QUIC channels will provide networking and mostly live
in the client library. Channel setup will require communication with the daemon. Current plans for language bindings
is to initially focus on the C API. The Rust client library is a bonus that we gain from the Rust -> C compilation. The C API 
should include autogenerated docs.

The daemon is able to participate in multiple teams using the same root identity by using some scheme
of generating leaf keybundles per team. This also removes the one-keybundle-per-process requirement.
For MVP we can lay the groundwork for the multi-keybundle approach, and better management of identities
can be added post-MVP or as time allows for MVP.

![daemon subsystems](/docs/images/new_daemon_arch.png)

![daemon subsystems with additional daemon](/docs/images/new_daemon_arch_detail.png)

## Config

On startup, the daemon requires a path to the working directory. That directory contains a configuration
file with the path to the unix domain socket that should be created and other values. The daemon working

> directory will structured to be compatible with the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/latest/). Adhering to the XDG Base Directory Specification will increase the interoperability of Aranya

Reviewer 4: The XDG scheme was suggested I think by Eric, I dont think we should keep it in $HOME, etc. Similar structure.

clients that have the full product integrated. AFC should also be compatible between subset implementations
and the full implementation. This compatibility is planned for Post-MVP.

##### AFC API 

The AFC API is being moved to a lower level in the API. They will still be available via a build flag to allow embedded devices and advanced users to access them. 

- `CreateAfcLabel(team_id, label)` - create a label
- `DeleteAfcLabel(team_id, label)` - delete a label
- `AssignAfcLabel(team_id, device_id, label)` - assign a label to a device so that it can be used for 
Quic Channels (or AFC)
- `RevokeAfcLabel(team_id, device_id, label)` - revoke a label from a device

> - `CreateBidiChannel(team_id, peer_net_ident, label) -> channel_id` - create a bidirectional channel with the given peer.

Author: ```suggestion
- `CreateAfcChannelBidi(team_id, peer_net_ident, label) -> channel_id` - create a bidirectional channel with the given peer.
```

- `DeleteChannel(team_id, channel_id)` - delete a channel.

> - `PollData(timeout)` - blocks until new AFC data is available, or timeout elapsed
> - `SendData(channel_id, data)` - send data on the given channel.
> - `RecvData() -> (data, metadata)` - recv data from AFC.

Author: ```suggestion
- `PollAfcData(timeout)` - blocks until new AFC data is available, or timeout elapsed
- `SendAfcData(channel_id, data)` - send data on the given channel.
- `ReceiveAfcData() -> (data, metadata)` - recv data from AFC.
```

Aranya Quic channels use a modified Quic transport implementation that supports the ability to use 
custom cryptography and has latency-based congestion control (see (s2n-quic)[https://github.com/aws/s2n-quic]).

Ideally, embedded devices that implement a subset of Aranya library should still be able to sync with
clients that have the full product integrated. AFC should also be compatible between subset implementations

> and the full implementation. This compatibility is Post-MVP.

Author: I think it's especially important to note that embedded devices will likely benefit more from AFC than Quic channels.

Reviewer 4: That is included in the initial paragraph in this section: " AFC is instead viewed as an experimental feature of the product which might still be preferred when running on embedded devices or when using a unidirectional transport."

Do you think it needs to be more clear?

Author: Nope, if it's included then that's probably enough

-  Fact DB queries via session commands
-  Standardize C API names to follow naming scheme.
-  Update key agreement commands to have a field for the sender's graph head ID. https://github.com/aranya-project/aranya-docs/pull/24#discussion_r1937642908
-  Update `AddTeam` to take a config

> -  Update `AddSyncPeer` to take a config

Author: Revert, these points are already listed on lines 369 and 371.
